package/gluon-core/files/lib/netifd/proto/gluon_wired.sh

```diff
@@ -7,15 +7,29 @@ init_proto "$@"
 proto_gluon_wired_init_config() {
         proto_config_add_boolean transitive
         proto_config_add_int index
+        proto_config_add_string vxpeer6addr
+        proto_config_add_boolean is_tunnel
 }
 
 xor2() {
         echo -n "${1:0:1}"
         echo -n "${1:1:1}" | tr '0123456789abcdef' '23016745ab89efcd'
 }
 
+is_layer3_device () {
+        local addrlen="$(cat "/sys/class/net/$1/addr_len")"
+        test "$addrlen" -eq 0
+}
+
 # shellcheck disable=SC2086
 interface_linklocal() {
+        if is_layer3_device "$1"; then
```

Comment:
What converts `mesh-vpn` (from https://github.com/freifunk-gluon/gluon/pull/2163/files#diff-44dd4d020d9ae0de557dc06405de0e7ce445eeb893d001cf1914a932f8bbedf6R11) to `mesh_vpn` needed for the ubus call?

Reply:
Netifd does that. `$ifname` (fetched from `$2`) in this script is **not** `ifname` from the uci section. `$ifname` is already resolved to `mesh_vpn` in this script.

I really dislike the naming conventions in netifd. It's just unnecessarily confusing. Scripts seem to name this argument `$2` differently:
```
# awk '/^proto_.*_setup/ { P=1; printf $1; } /^}/ {if (P) printf "\n"; P=0} /.*\$2.*/ { if(P) printf $0 }' /lib/netifd/proto/*
proto_batadv_setup()
proto_batadv_hardif_setup()	local iface="$2"
proto_batadv_vlan_setup()	local iface="$2"
proto_dhcp_setup()	local iface="$2"
proto_dhcpv6_setup()	local iface="$2"
proto_gluon_bat0_setup()
proto_gluon_mesh_setup()	export IFNAME="$2"
proto_gluon_wired_setup()        local ifname="$2"
proto_vxlan_setup()
proto_vxlan6_setup()
```

Naming it `iface` (like e.g. *proto_batadv_hardif_setup()*) seems to be better than naming it `ifname`. However still *proto_batadv_setup()* defines `iface="$config"` which means that `$iface=$1`:

```
root@lemoer-wgtest:/lib/netifd/proto# awk '/^proto_.*_setup/ { P=1; printf $1; } /^}/ {if (P) printf "\n"; P=0} /.*iface=.*/ { if(P) printf $0 }' /lib/netifd/proto/*
proto_batadv_setup()	local iface="$config"
proto_batadv_hardif_setup()	local iface="$2"
proto_batadv_vlan_setup()	local iface="$2"
proto_dhcp_setup()	local iface="$2"
proto_dhcpv6_setup()	local iface="$2"
proto_gluon_bat0_setup()
proto_gluon_mesh_setup()
proto_gluon_wired_setup()
proto_vxlan_setup()
proto_vxlan6_setup()
```

It's a mess in my oppinion. I don't know why netifd doesn't introduce something like "lower interface" or so. But maybe I am just not getting the concept...

Reply:
Yet the call in L19/L33 only works when `$1` is `mesh-vpn`.

```sh
# ubus call network.device status '{"name": "mesh_vpn"}'
Command failed: Invalid argument
```

vs.

```sh
# ubus call network.device status '{"name": "mesh-vpn"}'
{
	"external": false,
	"present": true,
	"type": "Network device",
	"up": true,
	"carrier": true,
	"link-advertising": [
		
	],
	"link-partner-advertising": [
		
	],
	"link-supported": [
		
	],
	"speed": "10F",
	"autoneg": false,
	"mtu": 1312,
	"mtu6": 1312,
	"macaddr": "42:b6:0e:6d:16:c7",
	"txqueuelen": 1000,
	"ipv6": true,
	"promisc": false,
	"rpfilter": 0,
	"acceptlocal": false,
	"igmpversion": 0,
	"mldversion": 0,
	"neigh4reachabletime": 30000,
	"neigh6reachabletime": 30000,
	"neigh4gcstaletime": 60,
	"neigh6gcstaletime": 60,
	"neigh4locktime": 100,
	"dadtransmits": 1,
	"multicast": true,
	"sendredirects": true,
	"statistics": {
		"collisions": 0,
		"rx_frame_errors": 0,
		"tx_compressed": 0,
		"multicast": 0,
		"rx_length_errors": 0,
		"tx_dropped": 0,
		"rx_bytes": 0,
		"rx_missed_errors": 0,
		"tx_errors": 0,
		"rx_compressed": 0,
		"rx_over_errors": 0,
		"tx_fifo_errors": 0,
		"rx_crc_errors": 0,
		"rx_packets": 0,
		"tx_heartbeat_errors": 0,
		"rx_dropped": 0,
		"tx_aborted_errors": 0,
		"tx_packets": 10560199,
		"rx_errors": 0,
		"tx_bytes": 1723304086,
		"tx_window_errors": 0,
		"rx_fifo_errors": 0,
		"tx_carrier_errors": 0
	}
}
```

Reply:
Ok, well you are right, I am wrong. `$2` to `proto_*_setup()` is acutally a device name, not an interface name (speaking in netifd terminology). So we have a problem here.

Reply:
```
ubus call network.interface dump | \
   jsonfilter -e "@.interface[@.l3_device='$1']['ipv6-address'][*].address" | \
   grep -e '^fe[89ab]' | head -n 1
```

Reply:
Applied this.